### PR TITLE
[MLX] C++ runner for HF LLMs with the off-graph KV cache

### PR DESCRIPTION
**Pull request by kiymetakdemir**
**Summary**

This runner builds an MLXSequenceCache, installs it, and passes the cache key, so it's the run path for .pte files exported with --use-offgraph-cache. The cache's shape is read from the .pte metadata. The flags left are policy the model can't imply: --kv-max-capacity, --kv-storage-dtype, --kv-initial-capacity, --kv-max-write, --kv-windows.

  Depends on #21680; the new CI job fails until that lands.

  **Files**

  - run_llm_hf.cpp — the runner: chat templates, greedy decode, benchmarking, and an interactive mode with /reset and /undo [N].
  - CMakeLists.txt — standalone find_package(executorch) project.
  - .github/workflows/mlx.yml — test-mlx-llm-offgraph for llama-1b, gemma3-1b and gemma4-e2b.

**Test**
CI builds the runner, then for llama-1b, gemma3-1b and gemma4-e2b exports off-graph and asserts the same "Paris" answer test-mlx-llm checks.

### CODE REVIEW

**Comment by metascroy** [comment]
Can you add something to the readme about this?

**Comment by metascroy** [comment]
@claude review this code

File: backends/mlx/examples/llm/run_llm_hf.cpp [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+// an MLXSequenceCache, installs it in the process-global registry, and passes
+// its cache_key as a load-time backend option (the rendezvous init() reads).
+// Its shape comes from constant methods the export publishes; the flags below
+// only choose policy. Without --kv-max-capacity it runs an in-graph model
+// unchanged -- so the same binary compares both cache paths. Greedy decode.
+//
+// Usage:
+//   run_llm_hf --pte <model.pte> --tokenizer <tokenizer.json> \
+//       [--kv-max-capacity N] [--kv-storage-dtype bf16|fp16|fp32] \
+//       [--kv-initial-capacity N] [--kv-max-write N] \
```
Comment (metascroy) [inline backends/mlx/examples/llm/run_llm_hf.cpp]: Why is kv-max-write a param?

Isn't it deriable from prefill-chunk-size?

File: backends/mlx/examples/llm/run_llm_hf.cpp [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+      return argv[i + 1];
+    }
+  }
+  return def;
+}
+
+} // namespace
+
+int main(int argc, char** argv) {
+  const std::string pte = arg(argc, argv, "--pte", "");
```
Comment (metascroy) [inline backends/mlx/examples/llm/run_llm_hf.cpp]: Can we use gflags?

File: backends/mlx/examples/llm/run_llm_hf.cpp [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+    // backend init reads its key, and the layout it needs is published by
+    // constant methods in the same file.
+    Module module(pte);
+    if (module.load() != Error::Ok) {
+      std::cerr << "Failed to load " << pte << std::endl;
+      return 1;
+    }
+
+    if (off_graph) {
+      cache::CacheConfig cfg{};
```
Comment (metascroy) [inline backends/mlx/examples/llm/run_llm_hf.cpp:378]: It's just a demo, but can we extract some of this cache config construction from pte metadata into a helper?

File: backends/mlx/examples/llm/run_llm_hf.cpp (hunk 1 above)
Comment (kiymetakdemir) [reply backends/mlx/examples/llm/run_llm_hf.cpp]: There was no chunking, but added --prefill-chunk-size (default 512, overrideable) and dropped --kv-max-write.

Comment (metascroy) [reply backends/mlx/examples/llm/run_llm_hf.cpp]: Yeah, we can add a --prefill-chunk-size to export script with default value of 512.

Write the chunk size to metadata that can be read in the runner.

Error out during export if chunk size is bigger than ring size on sliding wndow.

File: backends/mlx/examples/llm/run_llm_hf.cpp (hunk 2 above)
Comment (kiymetakdemir) [reply backends/mlx/examples/llm/run_llm_hf.cpp]: Changed to gflags.

File: backends/mlx/examples/llm/run_llm_hf.cpp (hunk 3 above)
Comment (kiymetakdemir) [reply backends/mlx/examples/llm/run_llm_hf.cpp:378]: Extracted.

File: backends/mlx/examples/llm/run_llm_hf.cpp [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+    cfg.layers[l].policy = w > 0
+        ? cache::LayerPolicy{cache::LayerPolicy::Kind::Ring, w}
+        : cache::LayerPolicy{cache::LayerPolicy::Kind::Flat, 0};
+  }
+  return cache::valid(cfg);
+}
+
+// argmax over the last position's vocab row of a [1, T, vocab] logits tensor,
+// reading whatever float dtype the op emitted.
+int64_t argmax_last(const ::executorch::aten::Tensor& logits) {
```
Comment (metascroy) [inline backends/mlx/examples/llm/run_llm_hf.cpp]: Don't we have a sampler/runner in ExecuTorch extension llm?

See llama runner and some of the other runners we have.

File: backends/mlx/examples/llm/run_llm_hf.cpp [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+      if (print_mem) {
+        // trailing space aligns the colon with the "after prefill" line above
+        print_footprint("after decode ", prompt_len + generated);
+      }
+
+      const double pf = ms(t0, t1), dc = ms(t1, t2);
+      const double pf_t = prompt_len / (pf / 1000.0);
+      const double dc_t = dc > 0 ? generated / (dc / 1000.0) : 0.0;
+      std::cout << "\n[iter " << iter << (measured ? "" : " warmup")
+                << "] prefill " << pf_t << " tok/s (" << prompt_len << " tok, "
```
Comment (metascroy) [inline backends/mlx/examples/llm/run_llm_hf.cpp]: I think we have something that collects stats on running?

Can you see some of the other C++ runners and reuse existign abstractions?

Comment (kiymetakdemir) [reply backends/mlx/examples/llm/run_llm_hf.cpp]: Switched to the shared Stats/print_report, removed --iters and changed --warmup to a bool to match the other runners and be able to use stats directly. Additionally adopted load_tokenizer and get_eos_ids abstractions.

File: backends/mlx/examples/llm/run_llm_hf.cpp (hunk 4 above)
Comment (kiymetakdemir) [reply backends/mlx/examples/llm/run_llm_hf.cpp]: Changed it to reuse extension.

File: backends/mlx/examples/llm/run_llm_hf.cpp (hunk 1 above)
Comment (kiymetakdemir) [reply backends/mlx/examples/llm/run_llm_hf.cpp]: Updated to this and if in runtime --kv-windows narrows window size it also narrows the chunk with it.

**Comment by kiymetakdemir** [comment]
The MLX LLM jobs skip on fork PRs (they need HF secrets), is there anything I can do to run those jobs?

**Comment by metascroy** [comment]
@claude review this code

**Comment by metascroy** [comment]
> The MLX LLM jobs skip on fork PRs (they need HF secrets), is there anything I can do to run those jobs?

You should be able to add the "ciflow/mlx" label to trigger the MLX jobs. I just added it.

**Review by metascroy — APPROVED** [review]

**Comment by kiymetakdemir** [comment]
I removed the kwargs check, HuggingFace forwards model-level arguments into attention (gemma 4 sends labels), so refusing anything unrecognised broke its export.
